### PR TITLE
nodeinit: Disable default ip-masq-agent jumps

### PR DESCRIPTION
**Pull request by dctrwatson**
Noticed some surprising behavior where disabling Cilium masquerade does not actually disable masquerade in GKE.

In GKE, even if the [IP masquerade agent](https://cloud.google.com/kubernetes-engine/docs/how-to/ip-masquerade-agent) is not enabled, the instance configure script [adds some default rules](https://github.com/kubernetes/kubernetes/blob/ae1103726f9aea1f9bbad1b215edfa47e0747dce/cluster/gce/gci/configure-helper.sh#L126-L143)

This removes the jump from `POSTROUTING` to the `IP-MASQ` chain

### CODE REVIEW

**Comment by dctrwatson** [comment]
Nevermind about AKS, I see AKS itself has added cilium as a supported network plugin provider that fixes this in the future. Just updating AKS doc can fix it now.

**Comment by nebril** [comment]
test-gke

**Review by aanm — COMMENTED** [review]
@tgraf can you take a look a this PR? Is it related with https://github.com/cilium/cilium/pull/11819?

File: install/kubernetes/cilium/charts/nodeinit/templates/daemonset.yaml [hunk 1]
```diff
@@ -81,6 +81,19 @@ spec:
                     systemctl restart kubelet
 {{- end }}
 
+{{- if .Values.global.gke.enabled }}
+                    # If the IP-MASQ chain exists, add back jump rule
```
Comment (tgraf) [inline install/kubernetes/cilium/charts/nodeinit/templates/daemonset.yaml]: This needs a more descriptive comment on why this is needed instead of what the code does. Why is this needed? Why does Cilium need to manipulate k8s iptables rules?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/charts/nodeinit/templates/daemonset.yaml (hunk 1 above)
Comment (dctrwatson) [reply install/kubernetes/cilium/charts/nodeinit/templates/daemonset.yaml]: @tgraf Add more info to the comment in the install script portion

**Comment by qmonnet** [comment]
test-gke

**Comment by nathanjsweet** [comment]
test-gke

**Comment by nathanjsweet** [comment]
retest-netnext

**Comment by nathanjsweet** [comment]
retest-4.19

**Comment by nathanjsweet** [comment]
retest-4.9

**Comment by nathanjsweet** [comment]
retest-net-next

**Comment by nathanjsweet** [comment]
retest-gke

**Review by tgraf — APPROVED** [review]

**Comment by nathanjsweet** [comment]
@dctrwatson I can't tell if you got another gke failure or not. I'm going to restart it again, but record what happened here.
[This is the gke output](https://jenkins.cilium.io/job/Cilium-PR-K8s-GKE/2061/artifact/) currently.

**Comment by nathanjsweet** [comment]
retest-gke

**Comment by dctrwatson** [comment]
I see this early on in the build:
```
09:42:37  FAIL: Cilium cannot be installed
09:42:37  Expected
09:42:37      <*helpers.cmdError | 0xc0006ae290>: Unable to generate YAML (exit status 1) output: cmd: helm template --validate /home/jenkins/workspace/Cilium-PR-K8s-GKE@2/src/github.com/cilium/cilium/install/kubernetes/cilium --namespace=cilium  --set global.cni.binPath=/home/kubernetes/bin  --set global.ipv6.enabled=false  --set global.hubble.enabled=true  --set global.tag=c991a94af799a9a774f7f8f8124746d956fdba10  --set global.hostFirewall=false  --set hubble-relay.image.tag=latest  --set hubble-relay.image.repository=147.75.55.179/cilium/hubble-relay:c991a94af799a9a774f7f8f8124746d956fdba10  --set operator.image=147.75.55.179/cilium/operator-generic:c991a94af799a9a774f7f8f8124746d956fdba10  --set global.debug.enabled=true  --set preflight.image=147.75.55.179/cilium/cilium:c991a94af799a9a774f7f8f8124746d956fdba10  --set global.k8s.requireIPv4PodCIDR=true  --set global.gke.enabled=true  --set global.nativeRoutingCIDR=10.0.0.0/8  --set config.ipam=kubernetes  --set global.bpf.preallocateMaps=true  --set global.etcd.leaseTTL=30s  --set config.bpfMasquerade=true  --set global.nodePort.mode=snat  --set global.ci.kubeCacheMutationDetector=true  --set global.logSystemLoad=true  --set config.enableCnpStatusUpdates=true  --set global.pprof.enabled=true  --set nodeinit.removeCbrBridge=true  --set nodeinit.reconfigureKubelet=true  --set agent.image=147.75.55.179/cilium/cilium:c991a94af799a9a774f7f8f8124746d956fdba10  --set global.hubble.listenAddress=:4244  --set global.ipv4.enabled=true  --set config.sessionAffinity=false  --set global.nodeinit.enabled=true  --set global.devices=  --set global.registry=147.75.55.179/cilium  > cilium-1626955e52c8bf98.yaml
09:42:37  Exitcode: 1 
09:42:37  Err: exit status 1
09:42:37  Stdout:
09:42:37   	 
09:42:37  Stderr:
09:42:37   	 Error: rendered manifests contain a resource that already exists. Unable to continue with install: existing resource conflict: kind: DaemonSet, namespace: cilium, name: cilium-node-init
```

**Comment by fristonio** [comment]
> I see this early on in the build:

I believe this is due to a [recent change](https://github.com/cilium/cilium/pull/12409/files#diff-2304891712e5d0140828499fc3a19043R3840-R3845) I introduced in tests while working on HA mode for `cilium-operator`. Can you point me to the link for these failures logs?
I think the fix should be pretty straightforward though, we can delete resources from the YAML we generated when removing cilium instead of manually removing each component. This will make sure we are removing all the deployed cilium components.

**Comment by dctrwatson** [comment]
> Can you point me to the link for these failures logs?

https://jenkins.cilium.io/job/Cilium-PR-K8s-GKE/2061/consoleFull

is where I pulled the excerpt from

**Comment by fristonio** [comment]
Sorry, but I think this has been fixed as a hotfix for GKE builds in #12725 
@nebril is currently working on some patches that will fix the remaining issues with GKE builds. We can retest once those are merged.

**Comment by fristonio** [comment]
retest-gke

**Comment by fristonio** [comment]
retest-gke
The cluster was not provisioned in previous run - https://jenkins.cilium.io/job/Cilium-PR-K8s-GKE/2131/

**Comment by dctrwatson** [comment]
Looks like just this over and over
```
07:41:17  Error from server (AlreadyExists): namespaces "cilium-ci-lock" already exists
07:41:18  
07:41:18  #15 [stage-3 2/9] COPY --from=builder /tmp/install /
07:41:18  #15 DONE 0.6s
07:41:18  Error from server (AlreadyExists): error when creating "/home/jenkins/workspace/Cilium-PR-K8s-GKE@2/src/github.com/cilium/cilium/test/gke/lock.yaml": deployments.apps "lock" already exists
07:41:18  error: --overwrite is false but found the following declared annotation(s): 'lock' already has a value (1)
07:41:18  1
[...]
07:41:23  Error from server (AlreadyExists): error when creating "/home/jenkins/workspace/Cilium-PR-K8s-GKE@2/src/github.com/cilium/cilium/test/gke/lock.yaml": deployments.apps "lock" already exists
```

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
@dctrwatson have you been able to manually test this in your own GKE environment?

The GKE CI job is not required, but this PR is specifically about GKE so the main thing we're looking for is reassurance that it works. We're working through some CI infrastructure provisioning issues with the GKE-based CI but as long as you've done due diligence this seems reasonable to merge.

**Comment by dctrwatson** [comment]
> @dctrwatson have you been able to manually test this in your own GKE environment?

Yes, have been using this with Cilium 1.7.x since a couple days before submitting this PR.

> EDIT: I should maybe add that seeing `connectivity-check` output that's successful with this patch would be a useful datapoint as a baseline.

GKE Node Version: `v1.16.13-gke.1`
Kernel Version: `4.19.112+`
Cilium Version: `1.7.7`

```

NAME                                                     READY   STATUS    RESTARTS   AGE
echo-a-58dd59998d-rk8w4                                  1/1     Running   0          72s
echo-b-865969889d-nw4qb                                  1/1     Running   0          71s
echo-b-host-659c674bb6-9qg6m                             1/1     Running   0          72s
host-to-b-multi-node-clusterip-6fb94d9df6-2nmw9          1/1     Running   0          71s
host-to-b-multi-node-headless-7c4ff79cd-k5ccz            1/1     Running   0          71s
pod-to-a-5c8dcf69f7-692j4                                1/1     Running   0          70s
pod-to-a-allowed-cnp-75684d58cc-kntqj                    1/1     Running   0          71s
pod-to-a-external-1111-669ccfb85f-m55hf                  1/1     Running   0          71s
pod-to-a-l3-denied-cnp-7b8bfcb66c-flfcr                  1/1     Running   0          70s
pod-to-b-intra-node-74997967f8-dlrlx                     1/1     Running   0          70s
pod-to-b-intra-node-nodeport-5fd7b6df5-kw566             1/1     Running   0          70s
pod-to-b-multi-node-clusterip-587678cbc4-5cq7q           1/1     Running   0          70s
pod-to-b-multi-node-headless-574d9f5894-t5cg8            1/1     Running   0          69s
pod-to-b-multi-node-nodeport-f544b6765-qhttw             1/1     Running   0          69s
pod-to-external-fqdn-allow-google-cnp-6dd57bc859-nz5jd   1/1     Running   0          69s
```

I used: https://raw.githubusercontent.com/cilium/cilium/v1.7/examples/kubernetes/connectivity-check/connectivity-check.yaml

Though I had to fix the port for the 2 nodeport pods:
```diff
--- connectivity-check.yaml	2020-08-07 18:45:53.970029385 -0700
+++ connectivity-check.yaml.orig	2020-08-07 18:47:32.146661105 -0700
@@ -365,13 +365,13 @@
         - -c
         - sleep 1000000000
         image: docker.io/byrnedo/alpine-curl:0.1.8
-        imagePullPolicy: IfNotPresent
+        imagePullPolicy: IfNotPresent 
         livenessProbe:
           exec:
-            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:41000" ]
+            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:31313" ]
         readinessProbe:
           exec:
-            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:41000" ]
+            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:31313" ]
         name: pod-to-b-intra-node-hostport
 ---
 apiVersion: apps/v1
@@ -506,10 +506,10 @@
         imagePullPolicy: IfNotPresent
         livenessProbe:
           exec:
-            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:41000" ]
+            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:31313" ]
         readinessProbe:
           exec:
-            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:41000" ]
+            command: [ "curl", "-sS", "--fail", "-o", "/dev/null", "echo-b-host-headless:31313" ]
         name: pod-to-b-multi-node-nodeport
```

**Comment by joestringer** [comment]
> Though I had to fix the port for the 2 nodeport pods:

Hmm. I don't actually see a nodeport declaration in that v1.7 version of the connectivity-check. The port 41000 seems to actually be a host application listening on the port. Seems like we need to fix that up. If you compare to the v1.8 YAML, there's actually a nodeport configured on port `31313` so no changes to the connectivity check should be necessary. Instead, we should extend the YAML to include the nodeport like this:

```
---
apiVersion: v1
kind: Service
metadata:
  name: echo-b
spec:
  type: NodePort
  ports:
  - port: 80
    nodePort: 31313
  selector:
    name: echo-b
```

**Comment by nebril** [comment]
test-gke

**Comment by dctrwatson** [comment]
I just tried with the v1.8 connectivity-check.yaml and it worked out of the box!

```
NAME                                                     READY   STATUS    RESTARTS   AGE
echo-a-58dd59998d-l7svd                                  1/1     Running   0          35s
echo-b-865969889d-hlxf4                                  1/1     Running   0          34s
echo-b-host-659c674bb6-mrm2z                             1/1     Running   0          35s
host-to-b-multi-node-clusterip-6fb94d9df6-hd5hv          1/1     Running   0          34s
host-to-b-multi-node-headless-7c4ff79cd-vjklg            1/1     Running   0          34s
pod-to-a-5c8dcf69f7-qhhhr                                1/1     Running   0          33s
pod-to-a-allowed-cnp-75684d58cc-zvd2k                    1/1     Running   0          34s
pod-to-a-external-1111-669ccfb85f-c9vh6                  1/1     Running   0          34s
pod-to-a-l3-denied-cnp-7b8bfcb66c-27fkr                  1/1     Running   0          33s
pod-to-b-intra-node-74997967f8-fkkkv                     1/1     Running   0          33s
pod-to-b-intra-node-nodeport-775f967f47-5ss46            1/1     Running   0          33s
pod-to-b-multi-node-clusterip-587678cbc4-hzlpc           1/1     Running   0          32s
pod-to-b-multi-node-headless-574d9f5894-86rnc            1/1     Running   0          32s
pod-to-b-multi-node-nodeport-7944d9f9fc-bbj7m            1/1     Running   0          32s
pod-to-external-fqdn-allow-google-cnp-6dd57bc859-nts9v   1/1     Running   0          32s
```

**Comment by nebril** [comment]
test-gke

**Comment by nebril** [comment]
The gke flake seems to be caused by a known flake, let's merge.

**Comment by joestringer** [comment]
Discussed during SIG-datapath meeting:

This fixes BPF masq (including ip-masq-agent) in v1.8. However the way it applies right now, if BPF masquerade is disabled, it could break ip-masq-agent configuration.

The default should be to enable BPF masquerade (on new enough kernels), in which case we want the logic introduced here. The concern we have is that when disabling BPF masquerade, this could cause unintended SNAT. The check switches on the `gke.enabled` helm flag right now, but it seems like there should also be a check for `config.bpfMasquerade` as well.

**Comment by dctrwatson** [comment]
Oh, interesting.

This was originally against v1.7, so `config.bpfMasquerade` didn't exist yet.

The intention is that by using Cilium, that ip-masq-agent is no longer used, thus making `global.masquerade` actually accurate.

I see 3 configurations:
1) `config.bpfMasquerade=false` + `global.masquerade=true`, use cilium iptables to masquerade
2) `config.bpfMasquerade=true` + `global.masquerade=true`, use cilium bpf to masquerade
3) `config.bpfMasquerade=false` + `global.masquerade=false`, no masquerading whatsoever

So I think another flag would have to added explicitly fallback to old behavior of using ip-masq-agent?

**Comment by Weil0ng** [comment]
FWIW, on GKE option 3 means we should just let GKE's ip-masq-agent do its thing, so the current behavior might be an issue in this case?

**Comment by joestringer** [comment]
@dctrwatson right now the check is just for whether GKE is enabled:

```
{{- if .Values.global.gke.enabled }}
```

Perhaps `global.masquerade`  would be the right thing to add here to address case #3?

**Comment by dctrwatson** [comment]
> FWIW, on GKE option 3 means we should just let GKE's ip-masq-agent do its thing, so the current behavior might be an issue in this case?

@Weil0ng that's the behavior that I found to be surprising and why I made this PR. My expectation was that Cilium manages all aspects of networking once installed by overriding kube-proxy, ip-masq-agent, etc (on new enough kernels.)

I'm not opposed to the old behavior of falling back to ip-masq-agent if docs are updated to reflect that in GKE, there are default masquerade rules that will always apply, and to actually disable masquerade to have 3 flags:
- `global.masquerade=false
- `global.gke.enabled=true`
- `global.gke.bypassIpMasqAgent=true` (this is the new flag, name subject to change)

> Perhaps a check for `global.masquerade` == false would be the right thing to add here to address case (3)?

@joestringer that would satisfy my intention for case (3). But as @Weil0ng points out, maybe not everyone wants to bypass ip-masq-agent rules?
